plugins/mac-capture/mac-screen-capture.m

```diff
@@ -0,0 +1,1049 @@
+
+		bool needs_to_update_properties = false;
+
+		if (!frame_detail_errored) {
+			if (sc->capture_type == ScreenCaptureWindowStream) {
+				if ((sc->frame.size.width !=
+				     window_rect.size.width) ||
+				    (sc->frame.size.height !=
+				     window_rect.size.height)) {
+					sc->frame.size.width =
+						window_rect.size.width;
+					sc->frame.size.height =
+						window_rect.size.height;
+					needs_to_update_properties = true;
+				}
+			} else {
+				size_t width =
+					CVPixelBufferGetWidth(image_buffer);
+				size_t height =
+					CVPixelBufferGetHeight(image_buffer);
+
+				if ((sc->frame.size.width != width) ||
+				    (sc->frame.size.height != height)) {
+					sc->frame.size.width = width;
+					sc->frame.size.height = height;
+					needs_to_update_properties = true;
+				}
+			}
+		}
+
+		if (needs_to_update_properties) {
+			[sc->stream_properties setWidth:sc->frame.size.width];
+			[sc->stream_properties setHeight:sc->frame.size.height];
+
+			[sc->disp
+				updateConfiguration:sc->stream_properties
+				  completionHandler:^(
+					  NSError *_Nullable error) {
+					  if (error) {
+						  MACCAP_ERR(
+							  "screen_stream_video_update: Failed to update stream properties with error %s\n",
+							  [[error localizedFailureReason]
+								  cStringUsingEncoding:
+									  NSUTF8StringEncoding]);
+					  }
+				  }];
+		}
+
+		prev_current = sc->current;
+		sc->current = frame_surface;
+		CFRetain(sc->current);
+		IOSurfaceIncrementUseCount(sc->current);
+
+		pthread_mutex_unlock(&sc->mutex);
+	}
+
+	if (prev_current) {
+		IOSurfaceDecrementUseCount(prev_current);
+		CFRelease(prev_current);
+	}
+}
+
+static inline void screen_stream_audio_update(struct screen_capture *sc,
+					      CMSampleBufferRef sample_buffer)
+{
+	CMFormatDescriptionRef format_description =
+		CMSampleBufferGetFormatDescription(sample_buffer);
+	const AudioStreamBasicDescription *audio_description =
+		CMAudioFormatDescriptionGetStreamBasicDescription(
+			format_description);
+
+	char *_Nullable bytes = NULL;
+	CMBlockBufferRef data_buffer =
+		CMSampleBufferGetDataBuffer(sample_buffer);
+	size_t data_buffer_length = CMBlockBufferGetDataLength(data_buffer);
+	CMBlockBufferGetDataPointer(data_buffer, 0, &data_buffer_length, NULL,
+				    &bytes);
+
+	CMTime presentation_time =
+		CMSampleBufferGetOutputPresentationTimeStamp(sample_buffer);
+
+	struct obs_source_audio audio_data = {};
+
+	for (uint32_t channel_idx = 0;
+	     channel_idx < audio_description->mChannelsPerFrame;
+	     ++channel_idx) {
+		uint32_t offset =
+			(uint32_t)(data_buffer_length /
+				   audio_description->mChannelsPerFrame) *
+			channel_idx;
+		audio_data.data[channel_idx] = (uint8_t *)bytes + offset;
+	}
+
+	audio_data.frames = (uint32_t)(data_buffer_length /
+				       audio_description->mBytesPerFrame /
+				       audio_description->mChannelsPerFrame);
+	audio_data.speakers = audio_description->mChannelsPerFrame;
```

Comment:
This will fail for 7 channels since we don't support it; check the enum there: https://github.com/obsproject/obs-studio/blob/master/libobs/media-io/audio-io.h#L67-L76

Reply:
Also a finer channel layout treatment would be a good idea. The https://github.com/obsproject/obs-studio/blob/master/libobs/media-io/audio-io.h#L67-L76 gives you the channel content for the layouts we support at obs. (And here is the correspondence to ffmpeg/avutil channel layouts : https://github.com/obsproject/obs-studio/blob/master/libobs/media-io/audio-resampler-ffmpeg.c#L69-L92 )
There could be a call to https://developer.apple.com/documentation/coremedia/1489137-cmaudioformatdescriptiongetchann?language=objc to retrieve the AudioChannelLayout and then throw an audio converter w/ kaudioconverterchannelmap.

Reply:
ScreenCaptureKit at this time only supports 1 or 2 channel audio, it's not possible to get 7 back in this situation. 

This is set to 2 in the configuration [here](https://github.com/obsproject/obs-studio/pull/6600/commits/6bd65159247466f7dfa0c990d0c62cac7e4e77e5#diff-461f0383ab3cdf344981fea3ddb7ac512e70cf0057edd6393118c6adf42d1447R417)

Reply:
> ScreenCaptureKit at this time only supports 1 or 2 channel audio, it's not possible to get 7 back in this situation.
> 
> This is set to 2 in the configuration [here](https://github.com/obsproject/obs-studio/pull/6600/commits/6bd65159247466f7dfa0c990d0c62cac7e4e77e5#diff-461f0383ab3cdf344981fea3ddb7ac512e70cf0057edd6393118c6adf42d1447R417)

ah fine then.
One question though, if the content has multichannel audio (ex: 5.1), how are these two channels set ?
- proper downmixing from 5.1 to stereo;
- or one just keeps the first two channels.
The latter would be very bad imo.
Second question, what about a mono source ? mikes are usually mono sources.

Reply:
I think it is downmix to stereo.

https://user-images.githubusercontent.com/17141779/174485322-18700898-28cb-46d3-8176-9e19879c8db3.mp4

Reply:
@pkviet does this sufficiently answer your concerns?

Reply:
Yeah fine. I just hope in the future that more channel layouts are supported than mono and stereo.